File: src/main/java/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.java
/*
 * Copyright (c) 2019 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.utils.buttons;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageReaction;
import net.dv8tion.jda.api.entities.TextChannel;
import org.cascadebot.cascadebot.CascadeBot;

import java.util.ArrayList;
import java.util.List;

@Getter
@RequiredArgsConstructor
public class ButtonGroup {

    private List<Button> buttons = new ArrayList<>();
    private final long ownerId;
    private final long channelId;
    private final long guildId;

    private long messageId = 0;

    private Runnable messageSentAction;

    public void addButton(Button button) {
        buttons.add(button);
        Guild guild = CascadeBot.INS.getShardManager().getGuildById(guildId);
        if (guild == null) return;
        TextChannel channel = guild.getTextChannelById(channelId);
        if (messageId != 0 && channel != null) {
            channel.retrieveMessageById(messageId).queue(button::addReaction);
        }
    }

    public void removeButton(Button button) {
        buttons.remove(button);
        Guild guild = CascadeBot.INS.getShardManager().getGuildById(guildId);
        if (guild == null) return;
        TextChannel channel = guild.getTextChannelById(channelId);
        if (messageId != 0 && channel != null) {
            channel.retrieveMessageById(messageId).queue(message -> {
                for (MessageReaction reaction : message.getReactions()) {
                    MessageReaction.ReactionEmote reactionEmote = reaction.getReactionEmote();
                    if (button instanceof Button.UnicodeButton && !reactionEmote.isEmote()) {
                        Button.UnicodeButton unicodeButton = (Button.UnicodeButton) button;
                        if (reactionEmote.getName().equals(unicodeButton.getUnicode())) {
                            reaction.removeReaction(CascadeBot.INS.getSelfUser()).queue();
                        }
                    } else if (button instanceof Button.EmoteButton && reactionEmote.isEmote()) {
                        Button.EmoteButton emoteButton = (Button.EmoteButton) button;
                        if (reactionEmote.getEmote().getIdLong() == emoteButton.getEmoteId()) {
                            reaction.removeReaction(CascadeBot.INS.getSelfUser()).queue();
                        }
                    }
                }
            });
        }
    }

    public Member getOwner() {
        return CascadeBot.INS.getShardManager().getGuildById(guildId).getMemberById(ownerId);
    }

    public void setMessageSentAction(Runnable messageSentAction) {
        this.messageSentAction = messageSentAction;
        if (messageId != 0) {
            this.messageSentAction.run();
        }
    }

    public void setMessage(long id) {
        messageId = id;
        if (messageSentAction != null) {
            messageSentAction.run();
        }
    }

    public void handleButton(Member clicker, TextChannel channel, Message buttonMessage, MessageReaction.ReactionEmote emote) {
        for (Button button : buttons) {
            if (button instanceof Button.EmoteButton && emote.isEmote()) {
                if (((Button.EmoteButton) button).getEmoteId() == emote.getEmote().getIdLong()) {
                    button.runnable.run(clicker, channel, buttonMessage);
                    return;
                }
            } else if (button instanceof Button.UnicodeButton && !emote.isEmote()) {
                if (((Button.UnicodeButton) button).getUnicode().equals(emote.getName())) {
                    button.runnable.run(clicker, channel, buttonMessage);
                    return;
                }
            }
        }
    }

    public void addButtonsToMessage(Message message) {
        if (buttons == null) return;
        for (Button button : buttons) {
            button.addReaction(message);
        }
        setMessage(message.getIdLong());
    }


}

File: src/main/kotlin/org/cascadebot/cascadebot/utils/buttons/ButtonGroup.kt
package org.cascadebot.cascadebot.utils.buttons

import de.bild.codec.annotations.Transient
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageReaction.ReactionEmote
import net.dv8tion.jda.api.entities.TextChannel
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.utils.buttons.Button.EmoteButton
import org.cascadebot.cascadebot.utils.buttons.Button.UnicodeButton
import java.util.*

open class ButtonGroup(
        val ownerId: Long,
        val channelId: Long,
        val guildId: Long
) {

    // Default constructor for MongoDB
    private constructor() : this(-1, -1, -1)

    @Transient
    private val buttons: MutableList<Button> = ArrayList()

    var messageId: Long = 0

    @Transient
    private var messageSentAction: Runnable? = null
    open fun addButton(button: Button) {
        buttons.add(button)
        val guild = CascadeBot.INS.shardManager.getGuildById(guildId) ?: return
        val channel = guild.getTextChannelById(channelId)
        if (messageId != 0L && channel != null) {
            channel.retrieveMessageById(messageId).queue { message: Message? -> message?.let{ button.addReaction(message)} }
        }
    }

    open fun removeButton(button: Button) {
        buttons.remove(button)
        val guild = CascadeBot.INS.shardManager.getGuildById(guildId) ?: return
        val channel = guild.getTextChannelById(channelId)
        if (messageId != 0L && channel != null) {
            channel.retrieveMessageById(messageId).queue { message: Message ->
                for (reaction in message.reactions) {
                    val reactionEmote = reaction.reactionEmote
                    if (button is UnicodeButton && !reactionEmote.isEmote) {
                        if (reactionEmote.name == button.unicode) {
                            reaction.removeReaction(CascadeBot.INS.selfUser).queue()
                        }
                    } else if (button is EmoteButton && reactionEmote.isEmote) {
                        if (reactionEmote.emote.idLong == button.emoteId) {
                            reaction.removeReaction(CascadeBot.INS.selfUser).queue()
                        }
                    }
                }
            }
        }
    }

    val owner: Member?
        get() = CascadeBot.INS.shardManager.getGuildById(guildId)!!.getMemberById(ownerId)

    fun setMessageSentAction(messageSentAction: Runnable?) {
        this.messageSentAction = messageSentAction
        if (messageId != 0L) {
            this.messageSentAction!!.run()
        }
    }

    fun setMessage(id: Long) {
        messageId = id
        if (messageSentAction != null) {
            messageSentAction!!.run()
        }
    }

    fun handleButton(clicker: Member?, channel: TextChannel?, buttonMessage: Message?, emote: ReactionEmote) {
        for (button in buttons) {
            if (button is EmoteButton && emote.isEmote) {
                if (button.emoteId == emote.emote.idLong) {
                    button.runnable.run(clicker, channel, buttonMessage)
                    return
                }
            } else if (button is UnicodeButton && !emote.isEmote) {
                if (button.unicode == emote.name) {
                    button.runnable.run(clicker, channel, buttonMessage)
                    return
                }
            }
        }
    }

    fun addButtonsToMessage(message: Message) {
        for (button in buttons) {
            button.addReaction(message)
        }
        setMessage(message.idLong)
    }

}
